File: src/main/ruby/g0001_0100/s0005_longest_palindromic_substring/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Dynamic_Programming
# #Data_Structure_II_Day_9_String #Algorithm_II_Day_14_Dynamic_Programming
# #Dynamic_Programming_I_Day_17 #Udemy_Strings #Big_O_Time_O(n)_Space_O(n)
# #2023_11_12_Time_117_ms_(98.63%)_Space_211_MB_(92.47%)

# @param {String} s
# @return {String}
def longest_palindrome(s)
  new_str = Array.new(s.length * 2 + 1, 0)
  new_str[0] = '#'

  (0...s.length).each do |i|
    new_str[2 * i + 1] = s[i]
    new_str[2 * i + 2] = '#'
  end

  dp = Array.new(new_str.length, 0)
  friend_center = 0
  friend_radius = 0
  lps_center = 0
  lps_radius = 0

  (0...new_str.length).each do |i|
    dp[i] = friend_center + friend_radius > i ? [dp[2 * friend_center - i], friend_center + friend_radius - i].min : 1

    while i + dp[i] < new_str.length && i - dp[i] >= 0 && new_str[i + dp[i]] == new_str[i - dp[i]]
      dp[i] += 1
    end

    if friend_center + friend_radius < i + dp[i]
      friend_center = i
      friend_radius = dp[i]
    end

    lps_center, lps_radius = i, dp[i] if lps_radius < dp[i]
  end

  s[(lps_center - lps_radius + 1) / 2, lps_radius - 1]
end

File: src/main/js/g0001_0100/s0005_longest_palindromic_substring/solution.js
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Dynamic_Programming
// #Data_Structure_II_Day_9_String #Algorithm_II_Day_14_Dynamic_Programming
// #Dynamic_Programming_I_Day_17 #Udemy_Strings #Big_O_Time_O(n)_Space_O(n)
// #2024_11_29_Time_10_ms_(99.54%)_Space_52.7_MB_(52.69%)

/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    const newStr = new Array(s.length * 2 + 1).fill('#')
    for (let i = 0; i < s.length; i++) {
        newStr[2 * i + 1] = s[i]
    }

    const dp = new Array(newStr.length).fill(0)
    let friendCenter = 0
    let friendRadius = 0
    let lpsCenter = 0
    let lpsRadius = 0

    for (let i = 0; i < newStr.length; i++) {
        dp[i] =
            friendCenter + friendRadius > i ? Math.min(dp[2 * friendCenter - i], friendCenter + friendRadius - i) : 1

        while (i + dp[i] < newStr.length && i - dp[i] >= 0 && newStr[i + dp[i]] === newStr[i - dp[i]]) {
            dp[i]++
        }

        if (friendCenter + friendRadius < i + dp[i]) {
            friendCenter = i
            friendRadius = dp[i]
        }

        if (lpsRadius < dp[i]) {
            lpsCenter = i
            lpsRadius = dp[i]
        }
    }

    const start = Math.floor((lpsCenter - lpsRadius + 1) / 2)
    const end = Math.floor((lpsCenter + lpsRadius - 1) / 2)
    return s.substring(start, end)
}

export { longestPalindrome }
